### PR TITLE
added wrapper for CBC functions

### PR DESCRIPTION
**Pull request by baraksalomon**

### CODE REVIEW

File: tests/test_tinyaes.py [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
 import tinyaes
 
+def pad(m):
+    str = chr(16-len(m)%16)*(16-len(m)%16)
+    return m+str.encode("utf-8")
```
Comment (naufraghi) [inline tests/test_tinyaes.py:10]: Is perhaps

```py
padding = bytes([16-len(m)%16]*(16-len(m)%16))
```

more idiomatic?

File: tests/test_tinyaes.py [hunk 2]
```diff
@@ -86,3 +111,16 @@ def test_different_keys_do_not_decode(data, aes_enc, aes_dec2, aes_dec3):
...
+    data = pad(data)
+    encoded = aes_enc_cbc.AES_CBC_encrypt_buffer(data)
+
+    assert aes_dec2_cbc.AES_CBC_decrypt_buffer(encoded) != data
+
+@given(data=binary(min_size=2, max_size=100))
+def test_cbc_decode(data, aes_enc2_cbc, aes_dec_cbc):
+    data = pad(data)
+    encoded = aes_enc2_cbc.AES_CBC_encrypt_buffer(data)
+    assert aes_dec_cbc.AES_CBC_decrypt_buffer(encoded) == data
```
Comment (naufraghi) [inline tests/test_tinyaes.py:126]: I'm not at all an expert in cryptography, but is `pad()` something like the `PKCS#7` algo suggested in the [tiny-AES-c/README](https://github.com/kokke/tiny-AES-c/tree/f06ac37fc31dfdaca2e0d9bec83f90d5663c319b#tiny-aes-in-c)?

I think in this case the raw binding could follow the wrapped library API, so your code is fine.

Thinking out loud, I'm not sure about the Python naming, should we now expose an `AES_CBC_encrypt_buffer_raw()` to make room for an opinionated `AES_CBC_encrypt_buffer()` that implements the suggested `PKCS#7` padding and decryption in `AES_CBC_decrypt_buffer()`? The ratio is that the main API exposed in Python might be correct and fast in most cases. The `raw` API is exposed for all other needs.

Given the _Important notes:_ in the original `README`, I expect that someone will misuse the API. The problem may be limited by more explicit naming.

Opinions?

**Review by naufraghi — COMMENTED** [review]
Thanks @baraksalomon for your PR, it seems nice, I just have one comment and I'm asking for an opinion.

File: tests/test_tinyaes.py (hunk 2 above)
Comment (baraksalomon) [reply tests/test_tinyaes.py:126]: pad() is an implementation of PKC#7.
As far as I know its the most commonly used padding for AES. its the default padding in OpenSSL.
AES-CTR works without padding. AWS-CBC must have padding. therefore for testing CBC padding must be used.
Regarding the naming, since this is a wrapper I think the APIs should as rule of thumb be named similarly to the wrapped API.
Regarding misusing the API, I think this can be addressed by documentation with examples.
having a fully functional minimal example in the documentation would provide a baseline to anyone who wants to use the APIs.
of course this is just an opinion, I'm not against an additional function that adds padding, but in that case I would suggest validating the padding and removing it in the decryption function.

File: tests/test_tinyaes.py (hunk 1 above)
Comment (baraksalomon) [reply tests/test_tinyaes.py:10]: Sure, I'll change that.

File: tests/test_tinyaes.py (hunk 2 above)
Comment (naufraghi) [reply tests/test_tinyaes.py:125]: Yes, but in fact we already made the distinction between `_inplace` and copy implementations.

I think a better API for this Python wrapper is to manage the padding internally when the plain `CBC_encrypt_buffer()` name is used.

TL;DR: can you turn your code into an implementation of the `_inplace` version of the methods? Maybe with a check on the length, if possible (see the other comment).

**Review by naufraghi — CHANGES_REQUESTED** [review]
Thanks for your feedback, but I prefer to keep the _plain_ `CTR_xcrypt_buffer` name for a future padded implementation.

Regarding your PR, I can merge a `CTR_xcrypt_buffer_inplace` implementation, everything else should be fine with minimal modifications.

I think the only thing missing is an update to the README. I'd just add a short note about the new methods and the padding requirement.

Let me know if you can handle that, or if you prefer to go for the minimal feature and let me update the README.

Thanks a lot!

File: tests/test_tinyaes.py (hunk 2 above)
Comment (baraksalomon) [reply tests/test_tinyaes.py]: changed the CBC functions to be `_inplace` and added block size check and test

File: tinyaes.pyx [hunk 3]
```diff
@@ -30,3 +30,12 @@ cdef class AES:
...
     def CTR_xcrypt_buffer_inplace(self, bytearray data):
         tinyaes.AES_CTR_xcrypt_buffer(&self._ctx, data, len(data))
+
+    def AES_CBC_encrypt_buffer(self, data):
+        inout = bytearray(data)
+        tinyaes.AES_CBC_encrypt_buffer(&self._ctx, inout, len(inout))
+        return bytes(inout)
+    def AES_CBC_decrypt_buffer(self, data):
+        inout = bytearray(data)
+        tinyaes.AES_CBC_decrypt_buffer(&self._ctx, inout, len(inout))
```
[the remark this answers is not in the document]
Comment (baraksalomon) [reply tinyaes.pyx]: added check for block size in CBC functions

**Comment by naufraghi** [comment]
Thanks @baraksalomon, I'm merging your PR.

I think I'll have to drop the Python 2 support in the next release because the CI is starting to complain. If that is the case, your feature will be in released in something more than `1.0.5`.

**Comment by baraksalomon** [comment]
Thanks!
